compiler/src/dotty/tools/dotc/transform/PatternMatcher.scala

```diff
@@ -332,11 +333,11 @@ object PatternMatcher {
           ref(defn.RuntimeTuplesModule)
             .select(defn.RuntimeTuples_apply)
             .appliedTo(receiver, Literal(Constant(i)))
-            .cast(args(i).tpe.widen)
 
         if (isSyntheticScala2Unapply(unapp.symbol) && caseAccessors.length == args.length)
           def tupleSel(sym: Symbol) = ref(scrutinee).select(sym)
-          val isGenericTuple = defn.isTupleClass(caseClass) && !defn.isTupleNType(tree.tpe)
+          val isGenericTuple = defn.isTupleClass(caseClass) &&
+            !defn.isTupleNType(tree.tpe match { case tp: OrType => tp.join case tp => tp }) // widen even hard unions, to see if it's a union of tuples
```

Comment:
I tried removing this change and all 3 tests passed. Maybe try removing it to see if something else breaks.

Was it there to generate `t._1` instead of `t(0)`?

If we do not need it, I would remove it.